### PR TITLE
Move access check to be external from Proxy

### PR DESCRIPTION
**Pull request by thodges-gh**
Summary of changes:
- Renames Whitelisted to Controller and to be abstract, forcing further implementation on inheriting contracts.
- Adds a SimpleAccessControl contract, which implements Controller, that enforces an access list updated by an owner.
- Renames WhitelistedAggregatorProxy to AccessControlledAggregatorProxy
- Updates AccessControlledAggregatorProxy to check an external Controller contract for users to be able to read functions.
- Updates AccessControlledAggregatorProxy to use a propose and confirm workflow for updating the aggregator.
- Updates the AggregatorProxy tests to use a MockAggregator to improve test time and clarity.

The general idea here is to have an aggregator proxy which should never need to be redeployed. This means aggregators and controller contracts can be updated independently of the proxy. In order to update an aggregator pointer, it must first be proposed, which allows users to verify that the new feed matches the current feed. The proposed aggregator must be confirmed in a separate transaction in order to update the pointer to the live aggregator. The controller contract can be queried to determine if the calling user has access, which puts the implementation of access on the external contract, rather than the proxy itself.

### CODE REVIEW

File: evm-contracts/src/v0.6/dev/ControllerInterface.sol [hunk 1]
```diff
@@ -0,0 +1,5 @@
+pragma solidity ^0.6.0;
+
+interface ControllerInterface {
+  function access(address user, bytes calldata data) external view returns (bool);
```
Comment (se3000) [inline evm-contracts/src/v0.6/dev/ControllerInterface.sol:4]: Doesn't this need to be public in order for checkAccess to call it in Controller.sol?

File: evm-contracts/src/v0.6/dev/AccessControl.sol [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  event CheckAccessEnabled();
+  event CheckAccessDisabled();
+
+  constructor()
+    public
+  {
+    checkEnabled = true;
+  }
+
+  function access(address _user) public view override returns (bool) {
```
[the remark this answers is not in the document]
Comment (thodges-gh) [reply evm-contracts/src/v0.6/dev/AccessControl.sol]: We don't want it on the actual access list because the access list could be a mapping of a mapping, which would change the signature. I'll add docs.

File: evm-contracts/src/v0.6/dev/Controller.sol [hunk 3]
```diff
@@ -0,0 +1,24 @@
+pragma solidity ^0.6.0;
+
+import "./ControllerInterface.sol";
+
+/**
+ * @title Controller
+ * @notice Allows inheriting contracts to control access to functions
+ */
+abstract contract Controller is ControllerInterface {
```
[the remark this answers is not in the document]
Comment (thodges-gh) [reply evm-contracts/src/v0.6/dev/Controller.sol:9]: I do like AccessController.

File: evm-contracts/src/v0.6/dev/ControllerInterface.sol (hunk 1 above)
Comment (thodges-gh) [reply evm-contracts/src/v0.6/dev/ControllerInterface.sol:4]: All functions declared in interfaces should be marked as external.

File: evm-contracts/src/v0.6/dev/ControllerInterface.sol [hunk 4]
```diff
@@ -0,0 +1,5 @@
+pragma solidity ^0.6.0;
+
+interface ControllerInterface {
+  function access(address user) external view returns (bool);
```
[the remark this answers is not in the document]
Comment (thodges-gh) [reply evm-contracts/src/v0.6/dev/ControllerInterface.sol]: `hasAccess` is good with me.

File: evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol [hunk 5]
```diff
@@ -105,10 +105,10 @@ contract WhitelistedAggregator is FluxAggregator, Whitelisted {
    * maxSubmissions) answer and updatedAt may change between queries.
    */
   function getRoundData(uint256 _roundId)
-    external
+    public
```
Comment (samsondav) [inline evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol:108]: For my own edification, why this change from `external` to `public`?

File: evm-contracts/src/v0.6/dev/AggregatorProxy.sol [hunk 6]
```diff
@@ -66,34 +68,34 @@ contract AggregatorProxy is AggregatorInterface, Owned {
...
-    external
+    public
     view
     virtual
     override
-    returns (uint256)
+    returns (uint256 roundId)
   {
-    return _latestRound();
+    ( roundId, , , , ) = latestRoundData();
```
Comment (samsondav) [inline evm-contracts/src/v0.6/dev/AggregatorProxy.sol:91]: Whoa solidity has pattern matching 🤯

File: evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol (hunk 5 above)
Comment (thodges-gh) [reply evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol:108]: Solidity made it easier to use functions marked as public in the implementation and external in the interface. This allowed us to get rid of the `_getRoundData(...)` internal functions which would be called by inheriting contracts and instead we can simply use `super.getRoundData(...)`.

File: evm-contracts/src/v0.6/tests/MockHistoricAggregator.sol [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import "../interfaces/AggregatorInterface.sol";
+
+/**
+ * @title The MockHistoricAggregator contract
+ * @notice Use this contract when you need to test
+ * other contract's ability to read data from an
+ * aggregator contract, but how the aggregator got
+ * its answer is unimportant
+ */
+contract MockHistoricAggregator is AggregatorInterface {
```
Comment (kaleofduty) [inline evm-contracts/src/v0.6/tests/MockHistoricAggregator.sol:12]: Should this be renamed to MockV2Aggregator?

File: evm-contracts/src/v0.6/dev/AccessController.sol [hunk 8]
```diff
@@ -0,0 +1,24 @@
+pragma solidity ^0.6.0;
+
+import "./AccessControllerInterface.sol";
+
+/**
+ * @title Controller
+ * @notice Allows inheriting contracts to control access to functions
+ */
+abstract contract AccessController is AccessControllerInterface {
```
Comment (kaleofduty) [inline evm-contracts/src/v0.6/dev/AccessController.sol:9]: Why do we need this contract?

File: evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol [hunk 9]
```diff
@@ -1,14 +1,14 @@
...
 
 /**
- * @title Whitelisted Prepaid Aggregator contract
- * @notice This contract requires addresses to be added to a whitelist
+ * @title AccessControlled FluxAggregator contract
+ * @notice This contract requires addresses to be added to a controller
  * in order to read the answers stored in the FluxAggregator contract
  */
-contract WhitelistedAggregator is FluxAggregator, Whitelisted {
+contract AccessControlledAggregator is FluxAggregator, SimpleAccessControl {
```
[the remark this answers is not in the document]
Comment (thodges-gh) [reply evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol:11]: I'd go even further and say it's externally controlled access control.

[the remark this answers is not in the document]
Comment (kaleofduty) [reply evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol:11]: works for me

[the remark this answers is not in the document]
Comment (kaleofduty) [reply evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol:11]: So `ExternallyAccessControlledAggregator`?

[the remark this answers is not in the document]
Comment (kaleofduty) [reply evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol:11]: @se3000 wdyt

[the remark this answers is not in the document]
Comment (se3000) [reply evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol:11]: Should we still signal that it's a proxy?

[the remark this answers is not in the document]
Comment (kaleofduty) [reply evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol:11]: This one isn't a proxy, is it?

File: evm-contracts/src/v0.6/dev/AccessController.sol (hunk 8 above)
Comment (thodges-gh) [reply evm-contracts/src/v0.6/dev/AccessController.sol:9]: It allows for a base contract that all inheriting contracts must create an implementation for `hasAccess` while providing a basic modifier to go along with it. I guess the argument for not needing this contract is that it doesn't have any implemented functions in it already. I'm open to removing it, but it's not necessary for the proxy itself.

File: evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (se3000) [reply evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol:11]: Right, but this one also does not have external access control. The proxy has the external access control, so the name could be `ExternallyAccessControlledAggregatorProxy`. Right? It's very verbose, would be nice to shorten it a bit but not necessary.

[the remark this answers is not in the document]
Comment (kaleofduty) [reply evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol:11]: Ah, right. Yes, the proxy would be `ExternallyAccessControlledAggregatorProxy`.

[the remark this answers is not in the document]
Comment (thodges-gh) [reply evm-contracts/src/v0.6/dev/AccessControlledAggregator.sol:11]: Ah right, at a glance I thought this comment was on the proxy.

File: evm-contracts/src/v0.6/dev/AccessController.sol (hunk 8 above)
Comment (kaleofduty) [reply evm-contracts/src/v0.6/dev/AccessController.sol:9]: My objection to it is that it doesn't seem to be used anywhere. 
SimpleAccessControl overrides hasAccess AND checkAccess.

If we want to keep it, checkAccess() should be made non-virtual.

File: evm-contracts/src/v0.6/dev/SimpleAccessControl.sol [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    public
+  {
+    checkEnabled = true;
+  }
+
+  /**
+   * @notice Returns the access of an address
+   * @param _user The address to query
+   */
+  function hasAccess(
```
Comment (kaleofduty) [inline evm-contracts/src/v0.6/dev/SimpleAccessControl.sol:30]: This function should also take into account checkEnabled.
Otherwise, there's no way to turn off access control for an EACAggregatorProxy without changing its accessController

File: evm-contracts/src/v0.6/interfaces/AggregatorInterface.sol [hunk 11]
```diff
@@ -1,29 +1,12 @@
 pragma solidity >=0.6.0;
 
-import "./HistoricAggregatorInterface.sol";
+interface AggregatorInterface {
```
Comment (kaleofduty) [inline evm-contracts/src/v0.6/interfaces/AggregatorInterface.sol:3]: I'm getting a bit confused by the naming of the interfaces. Isn't this an AggregatorV2Interface?

Comment (thodges-gh) [reply evm-contracts/src/v0.6/interfaces/AggregatorInterface.sol:3]: Can this be out of scope for this PR?

Comment (kaleofduty) [reply evm-contracts/src/v0.6/interfaces/AggregatorInterface.sol:3]: Ok

**Comment by kaleofduty** [comment]
LGTM apart from the above comments

File: evm-contracts/src/v0.6/dev/SimpleAccessControl.sol [hunk 12]
```diff
@@ -36,7 +36,7 @@ contract SimpleAccessControl is AccessControllerInterface, Owned {
     override
     returns (bool)
   {
-    return accessList[_user];
+    return checkEnabled ? accessList[_user] : true;
```
Comment (kaleofduty) [inline evm-contracts/src/v0.6/dev/SimpleAccessControl.sol:39]: Sorry to nitpick, but I think `accessList[_user]  || checkEnabled` would be cleaner

Comment (kaleofduty) [reply evm-contracts/src/v0.6/dev/SimpleAccessControl.sol:39]: Also, there's now a redundant || !checkEnabled in the checkAccess modifier.

And does checkAccess have to be virtual or could we get rid of that?

File: evm-contracts/src/v0.6/interfaces/AggregatorInterface.sol (hunk 11 above)
Comment (se3000) [reply evm-contracts/src/v0.6/interfaces/AggregatorInterface.sol:3]: I prefer leaving it as is, because then consumers right now don't have to change the paths they import things from. Alternatively, we could copy the file to AggregatorV2.sol, but I don't think we should remove the existing interface.
